clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2829,24 +2829,27 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 } // namespace
 
 class SYCLKernelNameTypeVisitor
-    : public TypeVisitor<SYCLKernelNameTypeVisitor>,
-      public ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor> {
+    : public TypeVisitor<SYCLKernelNameTypeVisitor, bool>,
+      public ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor, bool> {
   Sema &S;
   SourceLocation KernelInvocationFuncLoc;
-  using InnerTypeVisitor = TypeVisitor<SYCLKernelNameTypeVisitor>;
+  using InnerTypeVisitor = TypeVisitor<SYCLKernelNameTypeVisitor, bool>;
   using InnerTAVisitor =
-      ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor>;
+      ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor, bool>;
+  bool IsInvalid = false;
 
 public:
   SYCLKernelNameTypeVisitor(Sema &S, SourceLocation KernelInvocationFuncLoc)
       : S(S), KernelInvocationFuncLoc(KernelInvocationFuncLoc) {}
 
-  void Visit(QualType T) {
+  bool isValid() { return !IsInvalid; }
+
+  bool Visit(QualType T) {
     if (T.isNull())
-      return;
+      return false;
```

Comment:
So, we return `true` when everything is ok and false when something is bad. That means if we get null type or it is not a record, we think that everything is bad, but do not fire a diagnostic. Feels like these if conditions should be assertions instead.

Reply:
Tried changing this to assert and it causes a bunch of failures. I think it is possible that KernelNameType can be null and in such cases we return false instead of causing the program to abort.

Reply:
> I think it is possible that KernelNameType can be null and in such cases we return false instead of causing the program to abort.

Is it case when unnamed lambda is used?

Reply:
IIRC, if an unnamed lambda is passed as a kernel name, then the kernel name would be empty and we would get "kernel name is missing" diagnostic (if unnamedlambda is not supported)
```
QualType  KernelNameType =  calculateKernelNameType(getASTContext(), KernelFunc);
```
The null check here is to see if `KernelNameType`  from the above line will be null.
Also, when assert is used, it means that we assume `KernelNameType` can never be null. Whereas, when `if` is used, it implies that KernelNameType  could be null and we handle it by just returning. 
This has been my understanding.

Reply:
> Also, when assert is used, it means that we assume KernelNameType can never be null. 

I would expect that KernelNameType is null only if unnamed lambdas are enabled. Otherwise it is a class provided by the user or kernel object itself (in case when named callable object is used as kernel object).
Anyway, I think we can leave this for next patches.

Reply:
Ok. I will debug this case and confirm.